xarray/coding/variables.py

```diff
@@ -224,7 +224,7 @@ def _scale_offset_decoding(data, scale_factor, add_offset, dtype):
     return data
 
 
-def _choose_float_dtype(dtype, has_offset):
+def _choose_float_dtype_encoding(dtype, has_offset):
```

Comment:
Encoding per the CF spec is fairly specific. The packed variable is supposed to be type byte/short/int, not float. Most of the tests encode with a `scale_factor` or `add_offset` that require the packed data to be type float. Rather than trying to solve all this, I have just split the encode and decode dtype function.